java/ql/src/Security/CWE/CWE-094/ArbitraryAPKInstallation.ql

```diff
@@ -0,0 +1,129 @@
+ */
+
+import java
+import semmle.code.java.frameworks.android.Intent
+import semmle.code.java.dataflow.DataFlow
+import semmle.code.java.dataflow.TaintTracking2
+private import semmle.code.java.dataflow.ExternalFlow
+import DataFlow::PathGraph
+
+/** A string literal that represents the MIME type for Android APKs. */
+class PackageArchiveMimeTypeLiteral extends StringLiteral {
+  PackageArchiveMimeTypeLiteral() { this.getValue() = "application/vnd.android.package-archive" }
+}
+
+/** A method that sets the MIME type of an intent. */
+class SetTypeMethod extends Method {
+  SetTypeMethod() {
+    this.hasName(["setType", "setTypeAndNormalize"]) and
+    this.getDeclaringType() instanceof TypeIntent
+  }
+}
+
+/** A method that sets the data URI and the MIME type of an intent. */
+class SetDataAndTypeMethod extends Method {
+  SetDataAndTypeMethod() {
+    this.hasName(["setDataAndType", "setDataAndTypeAndNormalize"]) and
+    this.getDeclaringType() instanceof TypeIntent
+  }
+}
+
+/** A method that sets the data URI of an intent. */
+class SetDataMethod extends Method {
+  SetDataMethod() {
+    this.hasName(["setData", "setDataAndNormalize", "setDataAndType", "setDataAndTypeAndNormalize"]) and
+    this.getDeclaringType() instanceof TypeIntent
+  }
+}
+
+/** A dataflow sink for the URI of an intent. */
+class SetDataSink extends DataFlow::ExprNode {
+  SetDataSink() { this.getExpr().(MethodAccess).getMethod() instanceof SetDataMethod }
+}
+
+/** A method that generates a URI. */
+class UriConstructorMethod extends Method {
+  UriConstructorMethod() {
+    this.hasQualifiedName("android.net", "Uri", ["parse", "fromFile", "fromParts"]) or
+    this.hasQualifiedName("androidx.core.content", "FileProvider", "getUriForFile")
+  }
+}
+
+/**
+ * A dataflow source representing the URIs which an APK not controlled by the
+ * application may come from. Incuding external storage and web URLs.
+ */
+class ExternalAPKSource extends DataFlow::Node {
+  ExternalAPKSource() {
+    sourceNode(this, "android-external-storage-dir") or
+    this.asExpr().(MethodAccess).getMethod() instanceof UriConstructorMethod or
+    this.asExpr().(StringLiteral).getValue().matches(["file://%", "http://%", "https://%"])
+  }
+}
+
+/**
+ * A dataflow configuration for flow from an external source of an APK to the
+ * `setData[AndType][AndNormalize]` method of an intent.
+ */
+class APKConfiguration extends DataFlow::Configuration {
+  APKConfiguration() { this = "APKConfiguration" }
+
+  override predicate isSource(DataFlow::Node node) { node instanceof ExternalAPKSource }
+
+  override predicate isSink(DataFlow::Node node) {
+    exists(MethodAccess ma |
+      ma.getMethod() instanceof SetDataMethod and
+      ma.getArgument(0) = node.asExpr() and
+      any(PackageArchiveMimeTypeConfiguration c).hasFlowToExpr(ma)
+    )
+  }
+}
+
+/**
+ * A dataflow configuration tracking the flow of the Android APK MIME type to
+ * the `setType` or `setTypeAndNormalize` method of an intent.
```

Comment:
This isn't really true, is it? Because `setType[AndNormalize] are actually steps in this configuration, not sinks.

Also, this is now working because, coincidentally, there exists a summary that taints the return value of `setData` from its qualifier. But it would be much more readable to set the qualifier of `setData` as the sink of this configuration (the intent gets "tainted" from setting the MIME type, and then flows to the qualifier of `setData`, the argument of which is the sink of your other query).